gax-httpjson/src/main/java/com/google/api/gax/httpjson/ManagedHttpJsonChannel.java

```diff
@@ -43,62 +46,139 @@
-                  requestFactory.buildPostRequest(new GenericUrl(endpoint), content);
-              for (HttpJsonHeaderEnhancer enhancer : headerEnhancers) {
-                enhancer.enhance(request.getHeaders());
-              }
-              request.setParser(new JsonObjectParser(jsonFactory));
-              HttpResponse response = request.execute();
+              try (Writer stringWriter = new StringWriter()) {
+                GenericData tokenRequest = new GenericData();
+
+                // Create HTTP request body.
+                HttpRequestFormatter requestBuilder = methodDescriptor.httpRequestBuilder();
+                methodDescriptor.writeRequestBody(request, stringWriter);
+                stringWriter.close();
+                JsonHttpContent jsonHttpContent = null;
+                if (!Strings.isNullOrEmpty(stringWriter.toString())) {
+                  jsonFactory.createJsonParser(stringWriter.toString()).parse(tokenRequest);
+                  jsonHttpContent =
+                      new JsonHttpContent(jsonFactory, tokenRequest)
+                          .setMediaType((new HttpMediaType("application/json")));
+                }
+
+                // Populate HTTP path and query parameters.
+                Map<String, String> pathParams =
+                    requestBuilder.getPathParams(request, methodDescriptor.pathParams());
+                PathTemplate pathPattern =
+                    PathTemplate.create(methodDescriptor.endpointPathTemplate());
+                String relativePath = pathPattern.instantiate(pathParams);
+                GenericUrl url = new GenericUrl(endpoint + relativePath);
+                Map<String, List<String>> queryParams =
+                    requestBuilder.getQueryParams(request, methodDescriptor.queryParams());
+                for (String queryParam : methodDescriptor.queryParams()) {
+                  if (queryParams.containsKey(queryParam) && queryParams.get(queryParam) != null) {
+                    url.set(queryParam, queryParams.get(queryParam));
+                  }
+                }
+
+                HttpRequest httpRequest =
+                    requestFactory.buildRequest(
+                        methodDescriptor.httpMethod().name(), url, jsonHttpContent);
+
+                for (HttpJsonHeaderEnhancer enhancer : headerEnhancers) {
+                  enhancer.enhance(httpRequest.getHeaders());
+                }
+                httpRequest.setParser(new JsonObjectParser(jsonFactory));
 
-              GenericData responseData = response.parseAs(GenericData.class);
-              // TODO convert GenericData to response type
+                HttpResponse httpResponse = httpRequest.execute();
 
-              responseFuture.set(responseData);
+                ResponseT response =
+                    methodDescriptor.parseResponse(
+                        new InputStreamReader(httpResponse.getContent()));
+                responseFuture.set(response);
```

Comment:
The implementation under run() here should probably be put into its own class (I was too lazy to do this when bootstrapping)

Reply:
Done.